Modules/xASL_module_ASL.m

```diff
@@ -132,23 +144,147 @@
     x.DoWADQCDC = false; % default skip WAD-QC stuff
 end
 
-%% D3. Multi-PLD parsing
-if isfield(x.Q,'Initial_PLD') && numel(unique(x.Q.Initial_PLD))>1  % check for number of unique PLD's, more than 1 means multiPLD
-    fprintf(2,'Multi-PLD data detected, we will process this but this is a new feature that is still under development\n');
-    fprintf('PDLs: ');
-    for iPLD = 1:numel(x.Q.Initial_PLD)
-        if iPLD<numel(x.Q.Initial_PLD)
-            fprintf('%d, ',x.Q.Initial_PLD(iPLD));
-        else
-            fprintf('%d\n',x.Q.Initial_PLD(iPLD));
-        end
+
+%% D3. Multi-TE parsing
+% PM: we do 3 times the same below for TE, PLD, LabelingDuration respectively, we could refactor this code in a for-loop
+% PM: And perhaps we should move these to a separate function
+
+x.modules.asl.bMultiTE = false; % by default, we use single-echo processing
+% The echo-time is then only used to scale the M0 to the deltaM if they have different echo-times
+
+if ~isfield(x.Q, 'EchoTime')
+    warning('Missing field x.Q.EchoTime, this can be needed for quantification');
+    fprintf('%s\n', 'Defaulting to single-echo ASL processing');
+	x.Q.NumberEchoTimes = 1;
+elseif isempty(x.Q.EchoTime) || ~isnumeric(x.Q.EchoTime)
+    warning('Illegal field x.Q.EchoTime, this should not be empty and should contain numerical values');
+    fprintf('%s\n', 'Defaulting to single-echo ASL processing');
+	x.Q.NumberEchoTimes = 1;
+else
+    x.Q.uniqueEchoTimes = uniquetol(x.Q.EchoTime, 0.001);
+    x.Q.NumberEchoTimes = length(x.Q.uniqueEchoTimes); % Obtain the number of unique echo times
+    % PM: rename this parameter to x.Q.nUniqueTEs
+    
+    if sum(x.Q.EchoTime<=0)>0
+        warning('x.Q.EchoTime should be positive');
     end
-    x.modules.asl.bMultiPLD = true;
+
+    % Handle different number of echo times (== potentially different ASL sequences)
+    if x.Q.NumberEchoTimes==1
+        fprintf('%s\n', 'Single-echo ASL detected');
+    elseif x.Q.NumberEchoTimes==2
+        warning('Dual-echo ASL detected, dual-echo ASL processing not yet implemented');
+        fprintf('%s\n', 'We will process these ASL images as if they have a single echo, using the first echo only');
+        fprintf('%s\n', 'Assuming that this is a dual-echo ASL+BOLD sequence');
+        x.Q.EchoTime = min(x.Q.EchoTime);
+    elseif x.Q.NumberEchoTimes>2
+        x.modules.asl.bMultiTE = true;
+        fprintf('%s\n', 'Multiple echo times detected, processing this as multi-echo ASL');
+        fprintf('%s\n', 'Note that this feature is still under development');
+    end
+    
+    fprintf('%s\n', 'Detected the following unique echo times (ms):')
+    for iTE = 1:x.Q.NumberEchoTimes-1
+        fprintf('%.2f, ', round(x.Q.uniqueEchoTimes(iTE),4));
+    end
+    fprintf('%.2f\n', round(x.Q.uniqueEchoTimes(end),4));
+end
+
+% Deal with too short vectors (e.g., repetitions in the case of single-PLD)
+nTE = length(x.Q.EchoTime);
```

Comment:
A few redundant things in the code, plus you don't check if you can do the division. Without checking, repmat will potentially crash. Also, you a

Replace Lines 194-202 with the following code
nTE = length(x.Q.EchoTime);
if mod(nVolumes, nTE)
warning(['Number of echo times (n=' xASL_num2str(nTE) ') does not fit in the number of ASL volumes (n=' xASL_num2str(nVolumes) ')']);
else
factorTE = nVolumes/nTE;
x.Q.EchoTime = repmat(x.Q.EchoTime(:), [factorTE 1]);
end

Please do the same modifs for LabelingDuration and PLD, as the current code you have won't work properly.

Reply:
Yes will do

Reply:
Done in 0bdf54eb502121e43a25db8167411a91e183d051

Reply:
If you don't do the "else" the code will crash here already. 
Because nVolume/nTE won't be an integer. Repmat with non-integer factorTE will crash.

So if number of TEs doesn't fit in the number of volumes, then we have an invalid BIDS JSON and we can throw an error here.

Reply:
Indeed, we want it to crash here with a clear warning/error and not later with a vague unclear warning/error.
I thought of keeping the warning so that it is clear why it crashes with repmat. I have now converted this to a clear error: dc59f2a99a7cce88644bc78663b16cf426e823fa